### DISCUSSION TITLE
way to truncate tables after tests

### DISCUSSION THREAD

**Discussion by gavinking** [hibernate/hibernate-orm#3909] [Design Proposals]
I think we should add a schema export mode, `create-truncate`, or whatever, that deletes all the data without actually dropping the tables. This would be very helpful for testing on databases where DDL execution is slow.

(Of course there's other ways we could provide this functionality, but I'm not a fan of having a `truncate()` method on `Session`.)

**Comment by sebersole** [comment]
The one thing I do worry about is sequences.  Many databases do offer a means to reset them, but I doubt all do.  I guess worst case we just drop/create the sequences.

Also, the general interpretation for "create" in Hibernate is really "drop and create".  "create-drop" simply adds a drop at SF close also.  I only point this out because it is a deviation from the normal for that term and we need to be explicit about that - because "drop-create-truncate" is obviously pointless.

**Reply by gavinking** [reply]
And actually we don't even want to run the `create` statements every time I guess. Not sure what to do about that.

**Reply by sebersole** [reply]
Check for the existence ahead of time seems the only answer.  So `create-truncate` becomes "create the schema, if it does not exist, at boot and truncate data after".

Though another take... I can only see this as useful for tests; so maybe just handle this as a test-hook?  That's essentially what I had planned for ORM tests.

**Comment by gavinking** [comment]
Actually, do you folks think that a reason to *not* do this is that somebody will someday leave this setting turned on when deploying to production, and Hibernate's database user will have the permissions to do the `delete from` statements whereas it would not have had the permissions to execute DDL?

**Reply by gavinking** [reply]
I suppose the right thing to do would be to only expose this as a method of programmatic `SessionFactory` configuration, not as a config property.

**Reply by beikov** [reply]
Yes, I think it's much better to have a method on `SessionFactory` e.g. `truncate(boolean resetSequences)`

**Reply by Sanne** [reply]
I would not be over-protective of the user; if it's useful let's expose it. AFAIR `truncate` requires particular privileges, at least in some DBs, so we should be good on this front.

But I'm not super clear about the usefulness - I totally do remember using `truncate` as it's a way faster way to empty a large table, but in the context in which I was using it (large DBs with lots of tables and need for some seed data to be entered) I would have needed to use it "tacticaly" on selected tables only, and re-import some seed data in these tables only. Essentially only useful when a particular table needed to evolve the schema, trying to not modify all the other large tables.

Hibernate ORM could automate some aspects of such a smarter strategy but I don't think it can do it all. For example, it could validate the schema first to then only truncate + update the tables whose schema is not matching expectations. But we can't apply such smarter logic to our `import.sql` - we typically delegate the responsibility for more complex migrations to other tools.

But +1 to offer the option - I'm not at all against it, just not entirely sure of the workflow I'd use it in. Indeed for testing sounds great, like Steve mentioned.

**Reply by gavinking** [reply]
Well I don’t think `trúncate` is a very universal thing, so I think it’s going to be `delete` on most databases, right? Which you would have permission to do.

**Reply by beikov** [reply]
AFAICS the main use case for truncation is to avoid rebuilding the `SessionFactory` to get back to a clean database and putting that into a hbm2ddl option would kind of make this impossible I guess. Maybe we can add a filter parameter to that method to allow skipping truncation of certain tables for the seed data you mentioned.

**Reply by gavinking** [reply]
@beikov that’s a good point, we do want to avoid rebuilding the SF.

But actually the main motivation is that DDL execution is slow as fuck on some dbs, so I want to avoid it. Hibernate test suite spends most of its time on that, I believe.

**Reply by beikov** [reply]
>But actually the main motivation is that DDL execution is slow as fuck on some dbs, so I want to avoid it. Hibernate test suite spends most of its time on that, I believe.
>Well I don’t think truncate is a very universal thing, so I think it’s going to be delete on most databases, right? Which you would have permission to do.

Yeah, it's true that DDL is slow on some DBs, so +1 from me to avoid this as much as possible. You can look into our code base for implementations of the `org.hibernate.testing.cleaner.DatabaseCleaner` interface. There you will see that many database support truncation explicitly. Some DBs require that you disable foreign keys before deleting data though.

Thinking about this a bit more, maybe it's ok to have this just in our `hibernate-testing` module, since this is only relevant for testing. Wdyt?

**Reply by gavinking** [reply]
> Some DBs require that you disable foreign keys before deleting data though.

Yeah, I admit this is a problem. I was thinking of sorting the tables by foreign key direction, but, I dunno, a question I have simply never been able to answer in 20 years is: exactly how common is it to have circular FKs? How often does a parent really have the FK of one of its children? Because if it's not uncommon, then sorting the tables is not going to be that useful.

> Thinking about this a bit more, maybe it's ok to have this just in our `hibernate-testing` module

Well I wanted this for *users*. I doubt many of our users use `hibernate-testing`.

**Reply by gavinking** [reply]
Also: how many DBs allow turning off FK constraint checking (or anything equivalent)? I don't know the answer to this.

**Reply by beikov** [reply]
My point is, if this is only useful for testing (because why would anyone in production want to truncate a table?), then why not just advertise the `hibernate-testing` module a bit more since we already have this tooling there.
Disabling FK constraints is something that is used behind the scenes to implement truncation on certain DBs (currently DB2, SQL Server, Oracle, MySQL), but this should stay an implementation detail.

**Reply by gavinking** [reply]
> why not just advertise the hibernate-testing module a bit more since we already have this tooling there.

Well because `hibernate-testing` defines something that is more of a framework, whereas this is something that is useful even if you're *not* using that framework.

> DB2, SQL Server, Oracle, MySQL

Ah well, that's most of the ones I care about, actually. How does it work on Postgres?

**Reply by gavinking** [reply]
> You can look into our code base for implementations of the `org.hibernate.testing.cleaner.DatabaseCleaner` interface

Database cleaner looks interesting, but I'm not sure how I feel about the fact that it seems to trunc *all* tables in the schema, not just the ones mapped by Hibernate. Dunno. Perhaps that's fine.

**Reply by beikov** [reply]
>but I'm not sure how I feel about the fact that it seems to trunc all tables in the schema, not just the ones mapped by Hibernate

Yeah for our tests this was fine so far. 

If we want to expose this to users, we should move that to a dedicated module e.g. `hibernate-test-tools`? On PostgreSQL there is a `truncate` command which will defer FK checks to the end of the statement.

**Reply by gavinking** [reply]
Well, what I’m thinking now is to just give each dialect a `DatabaseCleaner` and then let you get it from the SF.

**Reply by beikov** [reply]
That sounds like a good idea.

**Reply by gavinking** [reply]
see https://github.com/hibernate/hibernate-orm/pull/5561

**Comment by gavinking** [comment]
OK, so this has grown into a whole `SchemaManager` API which I think looks kinda fine, and could grow over time.

What's there right now is still a little unfinished, lacking:

1. logging of the executed SQL (we need that!)
2. some refactoring of code duplication inherited from the original code in `hibernate-testing`
3. it doesn't properly support quoted identifiers
4. investigation into the purpose of some code which queried `sys.Tables` in the SQL server code, which I had to remove because that table was not accessible in the CI @beikov
5. maybe some handling of `table_type` column in `information_schema.tables`
6. support for Derby and Sybase (not a critical priority)

**Reply by gavinking** [reply]
Also, I'm uncertain whether I should be treating a MySQL database as a named catalog with a single schema which inherits the name of the catalog, or as an anonymous catalog with a single named schema. (Or some other variation on that.)

**Reply by sebersole** [reply]
I'm missing where you are proposing this `SchemaManager` contract.  What is its relation to `SchemaManagementTool`?

**Reply by gavinking** [reply]
Currently it's in a new package.

`SchemaManagementTool` is an internal thing that we run from a gradle task, or automatically when you set a system property. It's not intended to be invoked programmatically.

This is an API you can call directly in your program (usually, in a test).

**Reply by gavinking** [reply]
Also, it can do some things `SchemaManagementTool` can't do: truncate tables, for example.

**Reply by beikov** [reply]
AFAICT the usage of `sys.Tables` is only to be able to distinguish system tables from user tables i.e. the queries always use the predicate `is_ms_shipped = 0`.

**Reply by gavinking** [reply]
But is there some other way to do that here? A list of system schemas, for example?

**Reply by beikov** [reply]
That I don't know. I just wrote that stuff once so that it works and erased my memories afterwards 😆

**Comment by gavinking** [comment]
After all that, I ended up going back to my original idea of making this a part of the schema export framework, but only available via the new API.

And I ended up throwing away the code of `DatabaseCleaner`, and implementing something much simpler and more robust from scratch.

But I'm very happy with it now.
